### PR TITLE
[FIX] web: remove download_link widget + context on field widgets + enforce bin_size on binary

### PR DESCRIPTION
**Pull request by alexkuhn**
## 1. Remove download_link widget

Revert partially 255e8ca3d92e3a77d7c2a145677623518bc2f978

Removed widget `download_link`, as it was a sub-version of widget `binary`:
it did the exact same thing (representation of binary fields by means of a download link)
in a much better-looking way (download icon + filename as text of the download link,
instead of just a download link with text "Download").

Also, the values of binary fields might be already a human-readable binsize (e.g. 2.52 MBs),
instead of its plain textual string representation in base64:
The contextual parameter `bin_size`, when set to `true`, does not download the data of binary fields.
As a consequence, the server provides the sizes of the binary, instead of their value.

## 2. context on field widgets + enforce bin_size on binary

Field widgets now have a key `context` that let them
extend the context of the dataPoint (e.g. list).

In particular, widgets on binary fields enforce the contextual
item `{bin_size: True}`, so that the server gives the size of the
binary field as its value, instead of its content.

This change on binary fields reduces network load when accessing
a view with binary fields.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/js/core/utils.js [hunk 1]
```diff
@@ -23,6 +23,25 @@ var utils = {
             throw new Error("AssertionError");
         }
     },
+    /**
+     * Check if the value is a bin_size or not.
+     * If not, compute an approximate size out of the base64 encoded string.
+     *
+     * @param  {tring} value original format
```
Comment (Elkasitu) [inline addons/web/static/src/js/core/utils.js]: string?

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 2]
```diff
@@ -3069,13 +3069,23 @@ var BasicModel = AbstractModel.extend({
             id: {type: 'integer'},
         }, params.fields);
 
+        // datapoint context is extended by fields widgets
+        var context = params.context || {};
+        if (params.fieldsInfo && params.viewType) {
+            _.mapObject(params.fieldsInfo[params.viewType], function (field) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: _.each

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 3]
```diff
@@ -3069,13 +3069,23 @@ var BasicModel = AbstractModel.extend({
             id: {type: 'integer'},
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: missing docstring

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 4]
```diff
@@ -3069,13 +3069,23 @@ var BasicModel = AbstractModel.extend({
...
         }, params.fields);
 
+        // datapoint context is extended by fields widgets
+        var context = params.context || {};
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: var context = _.extend({}, params.context)

File: addons/web/static/src/js/core/utils.js [hunk 5]
```diff
@@ -23,6 +23,25 @@ var utils = {
...
     },
+    /**
+     * Check if the value is a bin_size or not.
+     * If not, compute an approximate size out of the base64 encoded string.
+     *
+     * @param  {string} value original format
+     * @return {string} bin_size (human-readable)
+     */
+    binaryToBinsize: function (value) {
+        if (!value) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/utils.js]: if the value can be something else than a string, then the signature is wrong

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 6]
```diff
@@ -3069,13 +3092,23 @@ var BasicModel = AbstractModel.extend({
             id: {type: 'integer'},
         }, params.fields);
 
+        // datapoint context is extended by fields widgets
+        var context = _.extend({}, params.context);
+        if (params.fieldsInfo && params.viewType) {
+            _.each(params.fieldsInfo[params.viewType], function (field) {
+                if (field.Widget && field.Widget.prototype.context) {
+                    context = _.extend(context, field.Widget.prototype.context);
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: this is the same as 
`_.extend(context, field.Widget.prototype.context);`

**Pull request merged**